vizro-core/src/vizro/models/_components/form/range_slider.py

```diff
@@ -54,32 +54,51 @@ def build(self):
             Output(self.id, "value"),
             Output(f"temp-store-range_slider-{self.id}", "data"),
         ]
-        input = [
+        inputs = [
             Input(f"{self.id}_start_value", "value"),
             Input(f"{self.id}_end_value", "value"),
             Input(self.id, "value"),
             State(f"temp-store-range_slider-{self.id}", "data"),
+            State(f"{self.id}_data", "data"),
         ]
 
-        @callback(output=output, inputs=input)
-        def update_slider_values(start, end, slider, input_store):
-            trigger_id = callback_context.triggered_id
-            if trigger_id == f"{self.id}_start_value" or trigger_id == f"{self.id}_end_value":
-                start_text_value, end_text_value = start, end
-            elif trigger_id == self.id:
-                start_text_value, end_text_value = slider
-            else:
-                start_text_value, end_text_value = input_store if input_store is not None else value
-
-            start_value = min(start_text_value, end_text_value)
-            end_value = max(start_text_value, end_text_value)
-            start_value = max(self.min, start_value)
-            end_value = min(self.max, end_value)
-            slider_value = [start_value, end_value]
-            return start_value, end_value, slider_value, (start_value, end_value)
+        clientside_callback(
```

Comment:
What's the advantage of moving these to a clientside_callback right now? Asking because currently in its current form it's very easy to unit test these callbacks as you can separate function from the calback. In the clientside_callback this seems more difficult to do now (See the latest added tests for sliders on this)

Reply:
In the current implementation, when the user changes the `Input ID_start_value`, a request is sent to the server and the server then calculates the new values for the `slider` and `Input ID_end_value` components. It took about 350ms and the filtering action was delayed by 350ms accordingly.

Implemented as pure client-side javascript code, this process is calculated in the user's browser and takes a few milliseconds to calculate.

---

> In the clientside_callback this seems more difficult to do now (See the latest added tests for sliders on this)

Good point! Let me see how we can test it.

Reply:
Two suggestions/thoughts that you may have considered already, but let's discuss (on the basis of info here: https://dash.plotly.com/clientside-callbacks):

- is it an option to use the assets folder? I assume not, but let's discuss
- have we considered a compromise for client-side callbacks:

As for the second, it could be advantageous to not go all the way with optimisations if it improves readability and testability. The say
> When the overhead cost of a callback becomes too great and no other optimization is possible, the callback can be modified to be run directly in the browser instead of a making a request to Dash.
which kind of hints that this is the last resort. Potential options I see:
- grouping the callbacks might help already a lot
- moving only some callbacks to client-side, and leaving the most complex code server side (have my eye slightly on the `function gateway` here)

Reply:
> is it an option to use the assets folder? I assume not, but let's discuss

Moving the javascript function into separate `static/js/` files looks much cleaner (functions become even more testable then).

> have we considered a compromise for client-side callbacks:
 
I considered it myself 😄. There are two reasons why the `gateway' callback should be implemented in javascript in my opinion:
   - Significant performance improvement. See `Speed-up' section in the PR description. If we implement the `gateway` callback as the server-side one, it would add about 400-800ms` to all the results in the `Optimised system` row. As a dashboard user, there is a big difference for me between 500ms and 1200ms waiting for filter results.
   - Another reason is that is much easier and more stable to implement in javascript cases such as (will be implemented next sprint): 
     - Handling handle when a new action chain is started before the previous one has finished.
     - Handling when number of actions in the queue (remaning_actions store) reached maximum limit (e.g. 100 actions).

Reply:
Could we move this to a `js` file then? I think then we can far more easily implement unit tests

Reply:
Let's continue conversation about unit testing [here](https://github.com/mckinsey/vizro/pull/34#discussion_r1329704237).